Block Selection Toolbar: Support fixed and sticky blocks by flipping when there is not enough space above the block

andrewserong opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Part of implementing: https://github.com/WordPress/gutenberg/issues/30121

An alternative to #46085.

This PR explores whether we can add support for fixed and sticky blocks to the block selection toolbar by flipping the position of the toolbar to be beneath the block when there is not enough space to display the toolbar above the block.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

<!--
Over in #46085 some of the limitations of the approach that switched on the `fixed` strategy were:

* The content area for the editor canvas has a variable height depending on the context. For example, there is the top toolbar present normally, however if the editor is not run in full-screen mode, then there is also the logged in wp-admin header. Retaining the `absolute` position means that we don't need to determine the clipping area for the popover, and can let floating UI figure that out for us.
* There was a jittery issue when switching over to `fixed` that was difficult to debug. That issue doesn't appear to be present with this PR, although I did notice there was still a bit of jitter on sticky/fixed blocks that might need investigating. -->

Without this change, blocks that are included inside a sticky positioned block at the top of the screen are obscured by the block selection toolbar. By flipping to display beneath the block when there is not enough room, we can ensure that the block content is not obscured.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

<!--
* In the block popover component, force a re-render whenever the scroll container for the editor canvas is scrolled. This forces the position of the popover to update when scrolling, which fixes positioning issues for fixed/sticky blocks. -->
* In the block toolbar popover props, ensure the toolbar will flip position to be underneath the block if there is not enough space to display the toolbar above the block when it is initially selected. This implements the feedback on the other PR (https://github.com/WordPress/gutenberg/pull/46085#issuecomment-1338937500):

> I think that in most cases we should not be flipping the position of the toolbar if it's able to start out above the block.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

To test the real world application for this PR, it might be easier to test #46142 instead, which includes the changes in this PR. However, you can still observe the changes in this PR with the following:

1. In the editor, select a block that begins well below the top of the visible area of the editor canvas. The block toolbar should be visible above the block. If you scroll the page so the block goes off the top of the viewport, then the toolbar should be "sticky" to the top of the editor canvas.
2. Select a block that is partially obscured at the top of the editor canvas. The block toolbar should be flipped and display beneath the block. As you scroll, the toolbar will flip between the top and bottom of the block. While this is a change from what's on `trunk`, it's this behaviour that ensures that for sticky and fixed position blocks, the toolbar will not obscure the content.

## Screenshots or screencast <!-- if applicable -->
| When a block is selected that does not have room for the toolbar initially (flips to bottom) | When a block is selected that does have enough room (does not flip) |
| --- | --- |
| <img width="867" alt="image" src="https://user-images.githubusercontent.com/14988353/207773538-3eba6718-e7ad-46eb-9a08-c1ba15f88a47.png"> | <img width="869" alt="image" src="https://user-images.githubusercontent.com/14988353/207773568-c3d0fe90-f107-45d4-b4c2-6792ee5013eb.png"> |

andrewserong committed:
Block Selection Toolbar: Support fixed and sticky blocks by re-rendering on scroll and flipping based on available space at top of viewport

andrewserong committed:
Add inline comments

talldan commented:
> In the block popover component, force a re-render whenever the scroll container for the editor canvas is scrolled. This forces the position of the popover to update when scrolling, which fixes positioning issues for fixed/sticky blocks.

I remember the popover used to have code that did this, but it was replaced with something else. This was the PR that added it - https://github.com/WordPress/gutenberg/pull/41402, this is the one that removed it - https://github.com/WordPress/gutenberg/pull/43617.

It'd be interesting to know a bit more about why it's needed again and why the change #43617 isn't sufficient.

> Select a block that is partially obscured at the top of the editor canvas. The block toolbar should be flipped and display beneath the block. As you scroll, the toolbar will flip between the top and bottom of the block. While this is a change from what's on trunk, it's this behaviour that ensures that for sticky and fixed position blocks, the toolbar will not obscure the content.

I found it difficult to reproduce this part, I'm not sure I'm doing it right. 😄

talldan commented on packages/block-editor/src/components/block-popover/index.js:
```diff
@@ -75,6 +76,34 @@ function BlockPopover(
 	}, [ selectedElement ] );
 
+	// Get the scrollable container that the block popover appears within.
+	const scrollContainer = useMemo( () => {
+		if ( ! __unstableContentRef?.current ) {
+			return;
+		}
+		return getScrollContainer( __unstableContentRef?.current );
+	}, [ __unstableContentRef?.current ] );
+
+	// Force the block popover to re-render whenever the content area is scrolled.
+	// This ensures that the position of the popover is accurate for fixed or sticky blocks.
+	useLayoutEffect( () => {
+		if ( ! scrollContainer ) {
+			return;
+		}
+
+		scrollContainer?.addEventListener?.(
+			'scroll',
+			forceRecomputePopoverDimensions
+		);
+
+		return () => {
+			scrollContainer?.removeEventHandler?.(
+				'scroll',
+				forceRecomputePopoverDimensions
+			);
+		};
+	}, [ scrollContainer ] );
```
Is this needed just for the toolbar? `BlockPopover` is used for a few different bits of UI.

`SelectedBlockPopover` is just for the selected block and handles the contextual toolbar, so it may be an optimisation to move the code there.

Though I don't think it has the `forceRecomputePopoverDimensions` function.

andrewserong replied:
It turns out I actually don't think we need any of the logic in the block popover component any more thanks to the changes that landed in https://github.com/WordPress/gutenberg/pull/46187 that resolved browser repaints of the popover when scrolling by switching transform values being passed to the animated wrapper. That's my hunch at least, but re-testing now against trunk, I can see this code isn't required now.

I'll re-do the description of this PR now that the scope of the change is a fair bit smaller (we just need to flipping logic now).

talldan commented on packages/block-editor/src/components/block-tools/use-block-toolbar-popover-props.js:
```diff
@@ -98,7 +131,12 @@ export default function useBlockToolbarPopoverProps( {
+				getProps(
+					contentElement,
+					selectedBlockElement,
+					scrollContainer,
```
`scrollContainer` should be in the `useCallback` dependencies.

andrewserong replied:
Good catch! Added.

andrewserong committed:
Add scrollContainer to deps array

andrewserong committed:
Remove no longer needed changes to block popover

andrewserong commented:
Thanks for testing and for the feedback @talldan!

After re-testing, it turns out that half of the changes in this PR were no longer required (the addition of the scroll event listener, etc) thanks to the improvements to the popover component that landed in https://github.com/WordPress/gutenberg/pull/46187 — that PR seems to have resolved the issues of the block popover not staying attached to the sticky/fixed element once scrolled off the page.

The remaining changes are still needed for sticky / fixed position support — that is, flipping the toolbar to be displayed below the block when there is initially not enough room to display it above the block. I've updated the PR description to reflect this smaller-scoped change. I've also updated #46142 (the sticky position support PR) to reflect the current state of this PR. For anyone testing, it might be simpler to play around with #46142 since the new flipping behaviour might be more visible there.

andrewserong commented:
I'll close out this PR now as its changed were merged as part of https://github.com/WordPress/gutenberg/pull/46142
